seed/data_importer/match.py

```diff
@@ -548,86 +559,91 @@ def states_to_views(unmatched_state_ids, org, access_level_instance, cycle, Stat
+    existing_views = ViewClass.objects.select_related("state").filter(cycle_id=cycle)
+    state_lookup: dict[int, Union[PropertyState, TaxLotState]] = {view.state.id: view.state for view in existing_views}
+    match_lookup: dict[tuple, list[dict[str, any]]] = defaultdict(list)
+    for state in ({k: getattr(view.state, k) for k in ["id", "hash_object", "updated", *matching_columns]} for view in existing_views):
+        match_lookup[tuple(state[c] for c in tuple_values)].append(state)
 
     if merge_duplicates:
-        duplicate_states = StateClass.objects.none()
+        duplicate_state_ids = set()
         duplicate_count = 0
     else:
         # Apply DATA_STATE_DELETE to incoming duplicate -States of existing -States in Cycle
         duplicate_states = StateClass.objects.filter(
-            pk__in=unmatched_state_ids, hash_object__in=Subquery(existing_states.values("hash_object"))
-        )
+            pk__in=unmatched_state_ids, hash_object__in=Subquery(existing_views.values("state__hash_object"))
+        ).only("id")
+        duplicate_state_ids = set(duplicate_states.values_list("id", flat=True))
         duplicate_count = duplicate_states.update(data_state=DATA_STATE_DELETE)
 
-    column_names = matching_criteria_column_names(org.id, table_name)
-
     # For the remaining incoming -States (filtering those duplicates), identify
     # -States with all matching criteria being None. These aren't eligible for matching.
-    empty_matching_criteria = empty_criteria_filter(StateClass, column_names)
-    promote_states = StateClass.objects.filter(pk__in=unmatched_state_ids, **empty_matching_criteria).exclude(
-        pk__in=Subquery(duplicate_states.values("id"))
+    empty_matching_criteria = empty_criteria_filter(StateClass, matching_columns)
+    promote_state_ids = set(
+        StateClass.objects.filter(pk__in=unmatched_state_ids.difference(duplicate_state_ids), **empty_matching_criteria).values_list(
+            "id", flat=True
+        )
     )
 
     # Identify and filter out -States that have been "handled".
-    handled_states = promote_states | duplicate_states
-    unmatched_states = StateClass.objects.filter(pk__in=unmatched_state_ids).exclude(pk__in=Subquery(handled_states.values("id")))
+    handled_state_ids = promote_state_ids.union(duplicate_state_ids)
+    unmatched_states = StateClass.objects.filter(pk__in=unmatched_state_ids.difference(handled_state_ids))
 
     # For the remaining -States, search for a match within the -States that are attached to -Views.
     # If one match is found, pass that along.
     # If multiple matches are found, merge them together, pass along the resulting record.
-    # Otherwise, add current -State to be promoted as is.
+    # Otherwise, add current -State to be promoted as-is.
     merged_between_existing_count = 0
-    merge_state_pairs = []
+    merge_state_pairs: list[Union[tuple[PropertyState, PropertyState], Union[TaxLotState, TaxLotState]]] = []
```

Comment:
I think that second Union should be a Tuple?

Reply:
Fixed